src/apps/subjects/api.py

```diff
@@ -296,3 +299,73 @@ async def get_my_subject(
             last_name=subject.last_name,
         )
     )
+
+
+async def get_target_subjects_by_respondent(
+    respondent_subject_id: uuid.UUID,
+    activity_or_flow_id: uuid.UUID,
+    user: User = Depends(get_current_user),
+    session: AsyncSession = Depends(get_session),
+    answer_session=Depends(get_answer_session),
+) -> ResponseMulti[TargetSubjectByRespondentResponse]:
+    subjects_service = SubjectsService(session, user.id)
+    respondent_subject = await subjects_service.get(respondent_subject_id)
+    if not respondent_subject:
+        raise NotFoundError(f"Subject with id {respondent_subject_id} not found")
+
+    if respondent_subject.user_id is None:
+        # Return a generic bad request error to avoid leaking information
+        raise ValidationError(f"Subject {respondent_subject_id} is not a valid respondent")
+
+    # Make sure the authenticated user has access to the subject
+    await CheckAccessService(session, user.id).check_subject_subject_access(
+        respondent_subject.applet_id, respondent_subject.id
+    )
+
+    assignment_service = ActivityAssignmentService(session)
+    assignment_subject_ids = await assignment_service.get_target_subject_ids_by_respondent(
+        respondent_subject_id=respondent_subject_id, activity_or_flow_ids=[activity_or_flow_id]
+    )
+
+    is_auto_assigned = await assignment_service.check_if_auto_assigned(activity_or_flow_id)
+    if is_auto_assigned:
+        assignment_subject_ids.append(respondent_subject_id)
+
+    submission_data: list[tuple[uuid.UUID, int]] = await AnswerService(
+        user_id=user.id, session=session, arbitrary_session=answer_session
+    ).get_target_subject_ids_by_respondent_and_activity_or_flow(
+        respondent_subject_id=respondent_subject_id, activity_or_flow_id=activity_or_flow_id
+    )
+
+    class SubjectInfo(TypedDict):
+        currently_assigned: bool
+        submission_count: int
+
+    subject_info: dict[uuid.UUID, SubjectInfo] = {}
+    for subject_id, submission_count in submission_data:
+        subject_info[subject_id] = {"currently_assigned": False, "submission_count": submission_count}
+
+    for subject_id in assignment_subject_ids:
+        if subject_id not in subject_info:
+            subject_info[subject_id] = {"currently_assigned": True, "submission_count": 0}
+        else:
+            subject_info[subject_id]["currently_assigned"] = True
+
+    subjects: list[Subject] = await subjects_service.get_by_ids(list(subject_info.keys()))
```

Comment:
I think my last (minor) request would be to ensure that the topmost subject returned in the array is the subject having the same ID as the respondent ID. It's just customary in the app to display self-reports first, followed by multi-informant reports. (No instruction has been provided by design for sorting the remaining items, so I'm not too concerned about that for now.)

Reply:
You got it